ir/program/program.rs

```diff
@@ -30,18 +40,238 @@ impl Program {
         &mut self.entry
     }
 
-    pub fn compile(match_: &typeql::query::stage::Match) -> Self {
-        let _entry = FunctionalBlock::from_match(match_);
-        todo!()
+    pub(crate) fn functions(&self) -> &Vec<FunctionIR> {
+        &self.functions
+    }
+
+    pub(crate) fn into_parts(self) -> (FunctionalBlock, Vec<FunctionIR>) {
+        let Self { entry, functions } = self;
+        (entry, functions)
+    }
+
+    pub fn compile<'index>(
+        function_index: &impl FunctionSignatureIndex,
```

Comment:
This now takes a single `FunctionSignatureIndex` instead of one having only the Schema in. It's clean, because it requires you to index even the preamble functions before calling it and separates out indexing and compilation completely.